### PR TITLE
Refactoring key generation functions

### PR DESCRIPTION
**Pull request by hilder-vitor**
Implemented function to encrypt polynomials into vector ciphertext.

Several parts of the key generation were actually encrypting polynomials
and integers into vector ciphertexts, thus, now they just call the new
function enc_ngs. As a result, the key generation code is simpler to
understand. Moreover, future applications may want to generate vector
ciphertexts, thus, the function enc_ngs is interesting on its own.

### CODE REVIEW

File: include/keygen.h [hunk 1]
```diff
@@ -122,4 +122,24 @@ class KeyGen
...
 
-#endif
\ No newline at end of file
+/**
+* Encrypts a polynomial into a vector ciphertext under the NTRU problem.
+* @param[out] ct ciphertext encrypting the input
+* @param[in] m polynomial to encrypt
+* @param[in] l dimension of the vector ciphertext
+* @param[in] B base used in the gadget vector
+* @param[in] sk_boot contains f and f^-1
```
Comment (iliailia) [inline include/keygen.h]: "contains bootstrapping secret key and its inverse"

File: include/keygen.h [hunk 2]
```diff
@@ -122,4 +122,24 @@ class KeyGen
...
+**/ 
+void enc_ngs(NGSFFTctxt& ct, const ModQPoly& m, int l, int B, const SKey_boot& sk_boot);
+
+/**
+* Encrypts an integer into a vector ciphertext under the NTRU problem.
+* @param[out] ct ciphertext encrypting the input
+* @param[in] m integer to encrypt (it is treated as a degree-0 polynomial)
+* @param[in] l dimension of the vector ciphertext
+* @param[in] B base used in the gadget vector
+* @param[in] sk_boot contains f and f^-1
```
Comment (iliailia) [inline include/keygen.h]: "contains bootstrapping secret key and its inverse"

File: src/keygen.cpp [hunk 3]
```diff
@@ -440,4 +279,58 @@ void KeyGen::get_bsk2(BSKey_LWE& bsk, const SKey_base_LWE& sk_base, const SKey_b
...
+    {
+        // sample random ternary vector
+        ModQPoly g(Param::N,0L);
+        Sampler::get_ternary_vector(g);
+        // FFT transform it
+        fftN.to_fft(g_fft, g);
+        // compute g * sk_boot^(-1)
+        tmp_ct = g_fft * sk_boot_inv_fft;
+        // compute g * sk_boot^(-1) + B^i * m
+        fftN.to_fft(msg_fft, msg); // msg = m * B^i
```
Comment (iliailia) [inline src/keygen.cpp]: This FFT can be done before the loop. Then in the loop, you can multiply `msg_fft` by `B` in the same way as you do here. See the code you removed above.

**Review by iliailia — CHANGES_REQUESTED** [review]
Thanks for the nice simplification. I left a few minor comments and one that should help to reduce the number of FFTs. 

Could you also indicate how this update changes the key generation time?

File: src/keygen.cpp (hunk 3 above)
Comment (hilder-vitor) [reply src/keygen.cpp]: Yes, that is true. I fixed it, thanks!

**Comment by hilder-vitor** [comment]
The bootstrapping key generations do not change much. For the NTRU, in average it took 0.73 s and now it takes 0.75 s (but generating the secret key of the base scheme already takes more than 10 s anyway) and for the LWE, it went from 0.211 s to 0.225 s in average.

**Review by iliailia — APPROVED** [review]
Thanks!

**Pull request merged**